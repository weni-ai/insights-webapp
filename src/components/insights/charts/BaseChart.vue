<template>
  <canvas
    ref="baseChartCanvas"
    :style="chartStyles"
  />
</template>

<script>
import {
  Chart as ChartJS,
  Title,
  Legend,
  BarElement,
  LineElement,
  PointElement,
  ArcElement,
  CategoryScale,
  LinearScale,
  RadialLinearScale,
  LineController,
  BarController,
} from 'chart.js';

import { deepMerge } from '@/utils/object';

const defaultPlugins = [
  Title,
  Legend,
  BarElement,
  LineElement,
  PointElement,
  ArcElement,
  CategoryScale,
  LinearScale,
  RadialLinearScale,
  LineController,
  BarController,
];

export default {
  name: 'BaseChart',
  props: {
    data: {
      type: Object,
      required: true,
    },
    options: {
      type: Object,
      default: () => ({}),
    },
    style: {
      type: Object,
      default: () => ({}),
    },
    type: {
      type: String,
      default: 'bar',
      validator: (value) =>
        [
          'bar',
          'line',
          'pie',
          'doughnut',
          'radar',
          'polarArea',
          'bubble',
          'scatter',
          'funnel',
        ].includes(value),
    },
    plugins: {
      type: Array,
      default: () => [],
    },
  },

  computed: {
    mergedOptions() {
      const defaultOptions = {
        maintainAspectRatio: false,
        responsive: true,
        barPercentage: 1.1,
        chart: {
          height: 100,
        },
        scales: {
          x: {
            grid: {
              display: false,
            },

            ticks: {
              padding: -2,
            },
          },
          y: {
            display: false,
          },
        },
        elements: {
          bar: {
            borderRadius: 4,
          },
        },
        plugins: {
          legend: {
            display: false,
          },
        },
      };
      return deepMerge(defaultOptions, this.options);
    },
    chartStyles() {
      const defaultStyles = {};
      return deepMerge(defaultStyles, this.style);
    },
  },

  created() {
    const pluginsToRegister = [...defaultPlugins, ...this.plugins];
    ChartJS.defaults.font.family = 'Lato, sans-serif';
    ChartJS.register(...pluginsToRegister);
  },

  mounted() {
    new ChartJS(this.$refs.baseChartCanvas, {
      type: this.type,
      data: this.data,
      options: this.mergedOptions,
    });
  },
};
</script>
