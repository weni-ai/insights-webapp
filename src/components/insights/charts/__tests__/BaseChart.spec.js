import { mount } from '@vue/test-utils';
import { describe } from 'vitest';

import BaseChart from '../BaseChart.vue';

describe('BaseChart', () => {
  it('Should match the snapshot', () => {
    const wrapper = mount(BaseChart);
    expect(wrapper.element).toMatchSnapshot();
  });
});
