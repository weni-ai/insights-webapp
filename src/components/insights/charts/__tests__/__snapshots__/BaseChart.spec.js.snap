// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseChart > Should match the snapshot 1`] = `
<canvas
  height="0"
  style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
  width="0"
/>
`;
